### DISCUSSION TITLE
EhcExecTransfer Timeout error

### DISCUSSION THREAD

**Discussion by vzg-certi** [tianocore/edk2#3430] [Q&A]
Hi. 
I'm developing an USB UEFI driver and my device is compatible only with USB 2HC protocol. I can set the configurations with all interfaces and get In and Out endpoint's but I everytime I try to send any packet with Bulk Transfer I receive Timeout error. 

The timeout error occures in `EhcExecTransfer`. I can create the URB (USB request Block) but when I try to send with EhcExecTransfer I receive a timeout error. 

I already try using Timeout = 0 but doesn't work the driver stays in a inifite loop trying to send the packet and returns nothing.

Someone can help me? I don't know if I have to do some handshake or something like that besides of the configurations that I already did.

**Comment by sambafruit** [comment]
Hi @vzg-certi ,
I encountered the similar issue when I read the data from the USB RNDIS Bulk In, but I am using XhcExecTransfer. I found the timeout is caused by inconsistent CycleBit of EventRingDequeue.

According to the USB spec, EventRingDequeue->CycleBit  should be the same as EventRingCCS, then this event is allowed to be processed. (`XhcSyncEventRing` function)

I haven't resolve this issue yet, I am still checking why Xhc doesn't update EventRingDequeue->CycleBit. :)
